parser/calcite/CalciteParserServer.java

```diff
@@ -81,24 +94,42 @@ public void parse(
         StreamObserver<CalciteParserProto.CalciteParserReply> responseObserver) {
 
       String q = request.getQuery();
-      Pair<Integer, SqlParseException> r = calcite(q);
-      if (r.getValue() == null) {
-        int i = r.getKey();
-        responseObserver.onNext(
-            CalciteParserProto.CalciteParserReply.newBuilder()
-                .setSql(q.substring(0, i))
-                .setExtension(q.substring(i))
-                .build());
-      } else {
-        responseObserver.onNext(
-            CalciteParserProto.CalciteParserReply.newBuilder()
-                .setError(r.getValue().getCause().getMessage())
-                .build());
-      }
+      Pair<Integer, SqlParseException> r = calciteParse(q);
+      int i = r.getKey();
+      String e = (r.getValue() == null ? "" : r.getValue().getCause().getMessage());
+      responseObserver.onNext(
+          CalciteParserProto.CalciteParserReply.newBuilder()
+          .setIndex(i)
+          .setError(e)
+          .build());
       responseObserver.onCompleted();
     }
 
-    private static int PosToIndex(String query, SqlParserPos pos) {
+    // calciteParse returns <len(query),null> if Calcite parser
+    // accepts query, or returns <pos,null> if a second parsing
+    // accepts the content to the left of the error position from the
+    // first parsing, or <-1,e> if both parsing failed.
+    private static Pair<Integer, SqlParseException> calciteParse(String query) {
+      try {
+        SqlParser parser = SqlParser.create(query);
+        SqlNode sqlNode = parser.parseQuery();
+
+      } catch (SqlParseException e) {
+        int epos = posToIndex(query, e.getPos());
+        try {
+          SqlParser parser = SqlParser.create(query.substring(0, epos));
+          SqlNode sqlNode = parser.parseQuery();
+        } catch (SqlParseException ee) {
+          return new Pair<Integer, SqlParseException>(epos, ee);
+        }
+        return new Pair<Integer, SqlParseException>(epos, null);
+      }
```

Comment:
If the `FROM` block is written at the end of the query, is the result of `calciteParse` correct?

`
SELECT *
TRAIN DNNClassifier
...
FROM table
`